lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs

```diff
@@ -16,39 +16,52 @@
 // under the License.
 
 using System;
-using System.Globalization;
 using System.Threading;
 using Org.Apache.REEF.Common.Tasks;
 using Org.Apache.REEF.Common.Tasks.Events;
-using Org.Apache.REEF.Examples.Tasks.HelloTask;
 using Org.Apache.REEF.Tang.Annotations;
 using Org.Apache.REEF.Utilities;
 using Org.Apache.REEF.Utilities.Diagnostics;
 using Org.Apache.REEF.Utilities.Logging;
 
 namespace Org.Apache.REEF.Tests.Functional.Messaging
 {
-    public class MessageTask : ITask, ITaskMessageSource
+    /// <summary>
+    /// A Task that sends messages to the Driver and receives messages from the
+    /// Driver for testing.
+    /// </summary>
+    public class MessageTask : ITask, ITaskMessageSource, IDriverMessageHandler
     {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(MessageTask));
+
         public const string MessageSend = "MESSAGE:TASK";
 
-        private static readonly Logger LOGGER = Logger.GetLogger(typeof(MessageTask));
+        // TODO[JIRA REEF-1385]: Check the MessageTaskSourceID on the Driver side.
+        public const string MessageTaskSourceId = "MessageTaskSourceId";
+
+        public const string MessageSentToDriverLog = "Message sent to Driver from Task.";
+        public const string MessageReceivedFromDriverLog = "Message received from Driver in Task.";
+
+        private readonly ManualResetEventSlim _messageFromDriverEvent = new ManualResetEventSlim(false);
+        private readonly MessageManager _messageManager;
 
         [Inject]
-        public MessageTask()
+        private MessageTask(MessageManager messageManager)
         {
+            _messageManager = messageManager;
         }
 
-        public HelloService Service { get; set; }
-
         public Optional<TaskMessage> Message
         {
             get
-            {              
-                TaskMessage defaultTaskMessage = TaskMessage.From(
-                    "messagingSourceId",
-                    ByteUtilities.StringToByteArrays(MessageSend + " generated at " + DateTime.Now.ToString(CultureInfo.InvariantCulture)));
-                LOGGER.Log(Level.Info, "Message is sent back from task to driver:" + defaultTaskMessage.Message);
+            {
+                var defaultTaskMessage = TaskMessage.From(
+                    MessageTaskSourceId,
+                    ByteUtilities.StringToByteArrays(MessageSend));
+                    
+                Logger.Log(Level.Info, MessageSentToDriverLog);
+                _messageManager.OnTaskMessageSent();
```

Comment:
You expect TaskMessage sent first, then release task, right? In this sequence, would it possible if task has completed but task message has not been received?